pep-0705.rst

```diff
@@ -0,0 +1,259 @@
+        if movie.get("year") is None:
+            return movie["name"]
+        else:
+            return f'{movie["name"]} ({movie["year"]})'
+
+Attempting to pass a ``MovieRecord`` into ``movie_string`` results (using mypy) in the error::
+
+    Argument 1 to "movie_string" has incompatible type "MovieRecord"; expected "Movie"
+
+This particular use case should be type-safe, but the type checker correctly stops the
+user from passing a ``MovieRecord`` into a ``Movie`` argument in the general case because
+the ``Movie`` class has mutator methods that could potentially allow the function to break
+the type constraints in ``MovieRecord`` by setting ``movie["year"] = None`` or ``del movie["year"]``.
+The problem disappears if we don't have mutator methods in ``Movie``. This could be achieved by defining an immutable interface using a :pep:`544` Protocol::
+
+    from typing import Literal, Protocol, overload
+
+    class Movie(Protocol):
+        @overload
+        def get(self, key: Literal["name"]) -> str: ...
+
+        @overload
+        def get(self, key: Literal["year"]) -> int | None: ...
+
+        @overload
+        def __getitem__(self, key: Literal["name"]) -> str: ...
+
+        @overload
+        def __getitem__(self, key: Literal["year"]) -> int | None: ...
+
+This is very repetitive, easy to get wrong, and is still missing important method definitions like ``__contains__()`` and ``keys()``.
+
+The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and support the cases exemplified above.
+In addition to those benefits, by flagging arguments of a function as ``TypedMapping``, it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desirable property of a function interface.
+Finally, this allows bringing the benefits of ``TypedDict`` to other mapping types that are unrelated to ``dict``.
+
+Specification
+=============
+
+The ``TypedMapping`` type is a protocol behaving almost identically to ``TypedDict``, except:
+
+1. The runtime type of a TypedMapping object is not constrained to be a ``dict``
+2. no mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated
+3. a class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases
+4. subclasses can narrow field types, in the same manner as other protocols
+
+All current and future features of TypedDict are applicable to TypedMapping, including class-based and alternative syntax, totality, and ``Required`` and ``NotRequired`` from :pep:`655`.
+
+As with :pep:`589`, this PEP provides a sketch of how a type checker is expected to support type checking operations involving TypedMapping and TypedDict objects, but details are left to implementors. In particular, type compatibility should be based on structural compatibility.
```

Comment:
Are there concrete reasons why we can't provide a full specification?

Reply:
Because it's incredibly complex and full of edge cases. For instance, type checkers are still working out how `@final` and `TypedDict` should interact (https://github.com/python/mypy/pull/13838). I think we'd take forever trying to nail a full spec down, and still get it wrong :(

Reply:
@JelleZijlstra ?

Reply:
Not entirely convinced. Type narrowing is indeed complicated, but that's different from providing a full specification of how type checking itself should work. In any case, this can be part of the conversation after the PEP is submitted.